### DISCUSSION TITLE
Aligning with a specific standard

### DISCUSSION THREAD

**Discussion by shunkica** [xmldom/xmldom#512] [General]
While working on adding JSDoc annotations to the `dom.js` module, I've observed a blend of DOM specifications being used - including XML DOM L2 Core, L3 Core, and the WHATWG Living Standard. There seems to be no discernible pattern behind this choice, which can make the library's behavior a bit unpredictable.

Given the name 'XMLDOM', adhering strictly to the XML DOM L3 Core spec might seem like the most logical choice. However, considering the library also claims to ponyfill for DOM in browsers, and recognizing that browsers are largely aligned with the Living Standard, there's a compelling argument for leaning in that direction too.

I'd love to hear everyone's thoughts on this, especially @karfau. Should we align more closely with L3 Core, or pivot towards the Living Standard to better mimic modern browser behavior?

For example:

We have the method `DOMImplementation.createHTMLDocument()`

This method is not part of any Level of XML DOM Core.

In [DOM L2 CORE](https://www.w3.org/TR/2000/CR-DOM-Level-2-20000510/html.html#ID-HTML-DOM) the function was introduced, but not on `DOMImplementation` but as a member of `HTMLDOMImplementation` which extends `DOMImplementation`

In the [living standard](https://dom.spec.whatwg.org/#dom-domimplementation-createdocument), the function is indeed present on the `DOMImplementation` interface

**Comment by karfau** [comment]
Thank you for taking the time to raise this question.

There is a related thread in https://github.com/xmldom/xmldom/issues/176 which mostly raises questions that should be considered along this topic, but doesn't come to any conclusion yet.

The readme states something related:
> When deciding how to fix bugs or implement features, xmldom tries to stay as close as possible to the various related specifications/standards.
> As indicated by the version starting with 0., this implementation is not feature complete and some implemented features differ from what the specifications describe.
> **Issues and PRs for such differences are always welcome, even when they only provide a failing test case.**

And the linked document regarding all the related specs gives a clear indication that the general decision is to at least document differences to tge living standard or even align the behavior, even if it means it is a breaking change:
https://github.com/xmldom/xmldom#dom

I would really like to find a way to run the relevant parts of the [web platform tests](https://github.com/web-platform-tests/wpt) using xmldom, to make it very obvious how aligned this library is with the browser implementation, but no time has been invested into how this would be possible. I just know that Deno seems to also be doing that, and that might be a good resource to look at.

Looking forward to any input/ideas on the topic.
It is just important to coordinate any breaking changes, and to keep PRs "reasonably small".

**Comment by shunkica** [comment]
Thank you for your input, @karfau. Your insights in discussion #176 are appreciated. However, I believe that the library name, xmldom, underscores a definitive commitment to using the XML DOM, and it should fully leverage the XML DOM Core.

Implementing and keeping up with the living standard seems unfeasible, given the scarcity of active contributors to this project. The HTML part seems to be added more as an after thought, barely even being mentioned in the projects README.

Of course, my viewpoint may be biased as I only utilize this library for handling XML documents.

**Reply by karfau** [reply]
If you look at the history of the project, the name was never intentionally chosen by the current maintainers, but inherited when forked and since we also inherited all the usages it doesn't make sense to limit the scope below what it already supports (to some degree).
The low amount of API endpoints currently offered by xmldom offer
- parsing of some XML and the HTML mime type, as is specified by the living standard into a DOM tree
- creating an HTML or XML DOM tree without parsing
- serializing a DOM tree back to a string

There are no plans to change that scope, since removing anything would make it fell even more incomplete.
(We might consider splitting it into multiple smaller packages to make it easier to compose only exactly what is needed, but xmldom would still be the thing that contains everything, just because that's how it has always been.)

The good news is that the changes in https://github.com/xmldom/xmldom/pull/338 which will be part of 0.9.0 and are currently available under the `next` tag, put the control of whether something is treated as XML or HTML into the code using our API (by making it mandatory to pass the mime type, and it can no longer be controlled by the content.
This makes it possible to consistently apply different rules for those mime types as demanded by the specs.

Regarding your contributions which I'm very happy about, it is totally fine if you focus on what you need to/want to, an it is fine for me to take care of that "manifold mindset".

I hope my answer helps :)

**Reply by shunkica** [reply]
A problem I encountered with the WHATWG standard is that the documentation is lacking, in the sense that there are no descriptions of the methods or parameters, just the implementation steps.
The XML DOM L2/L3 Core has much better documentation, but may diverge somewhat from the living standard.
Maybe we could take the documentation for JSDoc from the [Mozilla web api docs](https://developer.mozilla.org/en-US/docs/Web/API/Document_Object_Model) which should be aligned with WHATWG since Mozilla is a part of WHATWG.

**Reply by karfau** [reply]
Makes sense for me, and I think the current doc strings should at least be very close to it.
